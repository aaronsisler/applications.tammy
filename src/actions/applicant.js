import database from 'Firebase/firebase';
import {
    addApplicantNote,
    clearApplicant,
    setApplicant,
    setApplicantStatus,
    setApplicantNotes,
} from 'Actions/helpers/applicant';

export const startAddApplicantNote = (statusNote) => (dispatch, getState) => {
    const { applicantId, positionId } = getState().applicant;
    const date = new Date();
    const priority = -(date.getTime());

    return database.ref(`applicants/${positionId}/${applicantId}/applicantNotes`)
        .push().setWithPriority({ statusNote, priority }, priority)
        .then(() => {
            dispatch(addApplicantNote({ statusNote }));
        });
}

export const startClearApplicant = () => (dispatch) => dispatch(clearApplicant());

export const startSetApplicant = (applicantId) => (dispatch, getState) => {
    const { applicants } = getState();
    const applicantMatch = applicants.find((applicant) => applicant.applicantId == applicantId);

    return dispatch(setApplicant(applicantMatch));
}

export const startSetApplicantStatus = (applicantStatus) => (dispatch, getState) => {
    const { applicantId, positionId } = getState().applicant;

    return database.ref(`applicants/${positionId}/${applicantId}`).update({ applicantStatus })
        .then(() => dispatch(setApplicantStatus(applicantStatus)));
}

export const startSetApplicantNotes = () => (dispatch, getState) => {
    const { applicantId, positionId } = getState().applicant;
    return database.ref(`applicants/${positionId}/${applicantId}/applicantNotes`, {
        query: {
            orderByPriority: true,
        }
    }).then((snapshot) => {
        const applicantNotes = [];
        snapshot.forEach((childSnapshot) => applicantNotes.push(childSnapshot.statusNote));
        return dispatch(setApplicantNotes(applicantNotes))
    });
}
