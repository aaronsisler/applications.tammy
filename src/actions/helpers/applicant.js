export const addApplicantNote = (applicantNote) => ({
    type: 'ADD_APPLICANT_NOTE',
    applicantNote,
});

export const clearApplicant = () => ({
    type: 'CLEAR_APPLICANT',
});

export const setApplicant = (applicant) => ({
    type: 'SET_APPLICANT',
    applicant,
});

export const setApplicantStatus = (applicantStatus) => ({
    type: 'SET_APPLICANT_STATUS',
    applicantStatus,
});
