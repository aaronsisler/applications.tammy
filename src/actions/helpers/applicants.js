export const setApplicants = (applicants) => ({
    type: 'SET_APPLICANTS',
    applicants
});
