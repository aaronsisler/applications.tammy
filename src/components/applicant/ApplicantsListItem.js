import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

export class ApplicantsListItem extends React.Component {
    constructor(props) {
        super(props);
    }

    render() {
        return (
            <div className="applicants_list_item">
                <div className="positions_list_item__content">
                    <div className="positions_list_item__displayName">
                        {this.props.displayName}
                    </div>
                </div>
            </div>
        );
    }
}

/* istanbul ignore next */
const mapDispatchToProps = () => ({
});

export default connect(undefined, mapDispatchToProps)(ApplicantsListItem);

ApplicantsListItem.propTypes = {
    displayName: PropTypes.string,
};
