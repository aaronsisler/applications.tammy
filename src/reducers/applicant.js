const applicantReducerDefaultState = null;

const applicantReducer = (state = applicantReducerDefaultState, action) => {
    switch (action.type) {
        case 'SET_APPLICANT':
            return action.applicant;
        case 'CLEAR_APPLICANT':
            return null;
        case 'SET_APPLICANT_STATUS':
            return {
                ...state,
                applicantStatus: action.applicantStatus,
            }
        case 'ADD_APPLICANT_NOTE':
            return {
                ...state,
                applicantNotes: [action.applicantNote, ...state.applicantNotes],
            }
        default:
            return state;
    }
};

export default applicantReducer;

