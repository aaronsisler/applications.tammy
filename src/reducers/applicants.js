const applicantsReducerDefaultState = [];

const applicantsReducer = (state = applicantsReducerDefaultState, action) => {
    switch (action.type) {
        case 'SET_APPLICANTS':
            return action.applicants || applicantsReducerDefaultState;
        default:
            return state;
    }
};

export default applicantsReducer;

