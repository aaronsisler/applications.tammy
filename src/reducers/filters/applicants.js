const applicantsFilterReducerDefaultState =
{
    text: '',
};

const applicantsFilterReducer = (state = applicantsFilterReducerDefaultState, action) => {
    switch (action.type) {
        case 'SET_APPLICANTS_TEXT_FILTER':
            return {
                ...state,
                text: action.text
            };
        default:
            return state;
    }
};

export default applicantsFilterReducer;
