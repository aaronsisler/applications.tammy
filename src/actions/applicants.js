import database from 'Firebase/firebase';
import { setApplicants } from 'Actions/helpers/applicants';
import { setApplicantsTextFilter } from 'Actions/filters/applicantsText';

export const startSetApplicants = () => (dispatch, getState) => {
    const { positionId } = getState().workflow.position;

    return database.ref(`applicants/${positionId}`).once('value').then((snapshot) => {
        const applicants = [];
        snapshot.forEach((childSnapshot) => {
            const applicant = childSnapshot.val();
            const applicantNotes = applicant.applicantNotes
                ? [...Object.values(applicant.applicantNotes)]
                : [];
            applicants.push({
                applicantId: childSnapshot.key,
                positionId,
                ...childSnapshot.val(),
                applicantNotes,
            });
        });

        return dispatch(setApplicants(applicants));
    });
}

export const startSetApplicantsTextFilter = (text) => (dispatch) => dispatch(setApplicantsTextFilter(text));
