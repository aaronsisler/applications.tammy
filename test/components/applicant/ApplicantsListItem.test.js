import React from 'react';
import { shallow } from 'enzyme';
import { ApplicantsListItem, mapDispatchToProps } from 'Applicant/ApplicantsListItem';
import applicants from '../../fixtures/applicants';

import * as applicantActions from 'Actions/applicant';

describe('ApplicantsListItem', () => {
    const startSetApplicant = jest.fn();
    const [applicant] = applicants;
    let wrapper;

    beforeEach(() => {
        wrapper = shallow(
            <ApplicantsListItem
                applicantId={applicant.applicantId}
                startSetApplicant={startSetApplicant}
                {...applicant.user}
            />);
    });

    it('should render ApplicantsListItem correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should call startSetApplicant prop when clicked', () => {
        wrapper.find('.applicants_list_item').simulate('click');

        expect(startSetApplicant).toHaveBeenCalled();
    });

    describe('mapDispatchToProps', () => {
        it('should call startSetApplicant with applicantId', () => {
            const startSetApplicantMock = jest.spyOn(applicantActions, 'startSetApplicant');
            const dispatch = jest.fn();
            const { applicantId } = applicant;

            mapDispatchToProps(dispatch, { applicantId }).startSetApplicant();

            expect(startSetApplicantMock).toHaveBeenLastCalledWith(applicantId);
        });
    });
});
