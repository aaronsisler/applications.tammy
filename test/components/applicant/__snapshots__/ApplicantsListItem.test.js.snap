// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicantsListItem should render ApplicantsListItem correctly 1`] = `
<div
  className="applicants_list_item"
  onClick={[Function]}
>
  <div
    className="applicants_list_item__content"
  >
    <div
      className="applicants_list_item__display_name"
    />
    <div
      className="applicants_list_item__legal_name"
    >
      ,Â 
    </div>
  </div>
</div>
`;
